lxd/device/instance_id.go

```diff
@@ -0,0 +1,12 @@
+package device
+
+// instanceIdentifier is an interface that allows us to identify an Instance.
+type instanceIdentifier interface {
```

Comment:
@stgraber this new device package interface to represent a subset of the instance functionality allows us to remove the CanChangeLive() function from the Device interface as any devices that don't support hotplug can check IsRunning() on the instance passed in. It also will allow access to name and project of the instance, which is needed for bridged networking dnsmasq config generation.

I will also need to access network config for bridged devices, so was wondering how you would prefer this to be done, some approaches I was thinking about:

1. Pass daemon or daemon state into Device so that it can query directly the database and get the info it needs using functions like `networkLoadByName` with the daemon's state. Most flexible, but exposes lots to the Devices.
2. Detect bridge type network being requested inside c.deviceCreate and merge the network config properties into the device config properties (prefixed with either "network." or the "<network name>."). Much safer, but will require a device type specific logic in c.deviceCreate which we are trying to avoid.
3. Similar to 2, but imply the relationship between device's parent and managed network by checking if the parent property of the device can load a managed network config, if it can, then merge its config in as in 2. Avoids having device type specific logic in c.deviceCreate, but will make a call to `networkLoadByName` for every device that has a parent, even if not using a managed network, less efficient.

Reply:
Hmm, so one thing to be very careful with when giving access to `IsRunning()` is that any call to it will hang if called from within a liblxc hook. So using this means that you will not be able to use the device code from any of the On* functions at least.

Reply:
For the rest, I'd like to avoid special casing particular devices as much as possible but 1) may indeed be a bit too broad. I think passing in state.State is probably fine though and would allow access to the database functions.

If we find that to be too broad, then we could pass in a struct like:
```
type DeviceCallbacks struct {
    GetNetwork(name string) (*api.Network, error)
    GetStoragePool(name string) (*api.StoragePool, error)
    GetStoragePoolVolume(pool string, name string) (*api.StoragePoolVolume, error)
}
```

Exposing only the subset of information that we think is reasonable for devices to access and having those functions point to the same handlers we use to handle the matching `GET` requests.

Reply:
> Hmm, so one thing to be very careful with when giving access to `IsRunning()` is that any call to it will hang if called from within a liblxc hook. So using this means that you will not be able to use the device code from any of the On* functions at least.

Good point, I'd forgotten that, in which case I've remove IsRunning() from the instance interface and gone back to `CanHotPlug()` on the device itself, with a boolean passed to `deviceCreate` and `deviceStop` for "running".

Reply:
> For the rest, I'd like to avoid special casing particular devices as much as possible but 1) may indeed be a bit too broad. I think passing in state.State is probably fine though and would allow access to the database functions.
> 
> If we find that to be too broad, then we could pass in a struct like:
> 
> ```
> type DeviceCallbacks struct {
>     GetNetwork(name string) (*api.Network, error)
>     GetStoragePool(name string) (*api.StoragePool, error)
>     GetStoragePoolVolume(pool string, name string) (*api.StoragePoolVolume, error)
> }
> ```
> 
> Exposing only the subset of information that we think is reasonable for devices to access and having those functions point to the same handlers we use to handle the matching `GET` requests.

OK thanks, will go with state.State for now.